src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexOperationBase.cs

```diff
@@ -267,5 +274,73 @@ protected static int GetPageSize(IndexSearcher searcher, long pageSize)
 
             return (int)pageSize;
         }
+        
+        protected struct FilterDocument : IDisposable
+        {
+            private readonly IndexQueryServerSide _query;
+            private readonly DocumentsOperationContext _documentsContext;
+            private readonly Reference<int> _skippedResults;
+            private readonly Reference<int> _scannedDocuments;
+            private readonly IQueryResultRetriever _retriever;
+            private readonly QueryTimingsScope _queryTimings;
+            private readonly ScriptRunner.SingleRun _filterScriptRun;
+            private ScriptRunner.ReturnRun _filterSingleRun;
+            
+            public FilterDocument(Index index, IndexQueryServerSide query, DocumentsOperationContext documentsContext, Reference<int> skippedResults, Reference<int> scannedDocuments, IQueryResultRetriever retriever,QueryTimingsScope queryTimings)
+            {
+                _query = query;
+                _documentsContext = documentsContext;
+                _skippedResults = skippedResults;
+                _scannedDocuments = scannedDocuments;
+                _retriever = retriever;
+                _queryTimings = queryTimings;
+                if (query.Metadata.FilterScript is not null)
+                {
+                    var key = new CollectionQueryEnumerable.FilterKey(query.Metadata);
+                    _filterSingleRun = index.DocumentDatabase.Scripts.GetScriptRunner(key, readOnly: true, patchRun: out _filterScriptRun);
+                }
+                else
+                {
+                    _filterScriptRun = null;
+                    _filterSingleRun = default;
+                }
+            }
+
+            public bool TryFilter(global::Lucene.Net.Documents.Document document, string key, IState state, out bool overLimit)
```

Comment:
Better to return an enum with `Value / Skip / Done` - otherwise, you have `if(overLimit) break; continue;` everywhere

Reply:
Done.